src/mediastream.cc

```diff
@@ -0,0 +1,270 @@
+
+  MediaStream* obj = new MediaStream(msi);
+  msi->RegisterObserver(obj);
+  obj->Wrap( info.This() );
+
+  TRACE_END;
+  info.GetReturnValue().Set( info.This() );
+}
+
+void MediaStream::QueueEvent(AsyncEventType type, void* data)
+{
+  TRACE_CALL;
+  AsyncEvent evt;
+  evt.type = type;
+  evt.data = data;
+  uv_mutex_lock(&lock);
+  _events.push(evt);
+  uv_mutex_unlock(&lock);
+
+  uv_async_send(&async);
+  TRACE_END;
+}
+
+void MediaStream::Run(uv_async_t* handle, int status)
+{
+  TRACE_CALL;
+
+  Nan::HandleScope scope;
+
+  MediaStream* self = static_cast<MediaStream*>(handle->data);
+  TRACE_CALL_P((uintptr_t)self);
+  //Local<Object> ms = self->handle();
+  bool do_shutdown = false;
+
+  while(true)
+  {
+    uv_mutex_lock(&self->lock);
+    bool empty = self->_events.empty();
+    if(empty)
+    {
+      uv_mutex_unlock(&self->lock);
+      break;
+    }
+    AsyncEvent evt = self->_events.front();
+    self->_events.pop();
+    uv_mutex_unlock(&self->lock);
+
+    TRACE_U("evt.type", evt.type);
+    if(MediaStream::CHANGE & evt.type)
+    {
+
+    }
+    
+    if(MediaStream::ACTIVE & evt.type) 
+    {
+
+    } else if(MediaStream::INACTIVE & evt.type) 
+    {
+
+    }
+    if(MediaStream::ADDTRACK & evt.type) 
+    {
+
+    }
+    if(MediaStream::REMOVETRACK & evt.type) 
+    {
+
+    }
+  }
+
+  if (do_shutdown) {
+    uv_close(reinterpret_cast<uv_handle_t*>(&self->async), nullptr);
+  }
+
+  TRACE_END;
+}
+
+void MediaStream::OnChanged()
+{
+  TRACE_CALL;
+  QueueEvent(MediaStream::CHANGE, static_cast<void*>(NULL));
+  TRACE_END;
+}
+
+NAN_METHOD(MediaStream::GetAudioTracks) {
+  TRACE_CALL;
+  MediaStream* self = ObjectWrap::Unwrap<MediaStream>( info.Holder() );
+
+  auto audioTracks = self->GetInterface()->GetAudioTracks();
+  auto audioTracksJSArray = Nan::New<v8::Array>(audioTracks.size());
+  int i = 0;
+  for(auto track : audioTracks) {
+    auto jsAudioTrack = Nan::New<v8::Object>();
+    Nan::Set(jsAudioTrack, Nan::New<v8::String>("id").ToLocalChecked(),
+             Nan::New<v8::String>(track->id()).ToLocalChecked());
+
+    Nan::Set(jsAudioTrack, Nan::New<v8::String>("kind").ToLocalChecked(),
+             Nan::New<v8::String>(track->kind()).ToLocalChecked());
+
+    Nan::Set(jsAudioTrack, Nan::New<v8::String>("label").ToLocalChecked(),
+             Nan::New<v8::String>("").ToLocalChecked());
+
+    Nan::Set(jsAudioTrack, Nan::New<v8::String>("muted").ToLocalChecked(),
+             Nan::New<v8::Boolean>(track->GetSource()->state()
+                                   == track->GetSource()->kMuted));
+
+    // ToDO: Determine if track is read only.
+    Nan::Set(jsAudioTrack, Nan::New<v8::String>("readonly").ToLocalChecked(),
+             Nan::New<v8::Boolean>(true));
+
+    Nan::Set(jsAudioTrack, Nan::New<v8::String>("remote").ToLocalChecked(),
+             Nan::New<v8::Boolean>(track->GetSource()->remote()));
+
+    Nan::Set(audioTracksJSArray, Nan::New<Number>(i++), jsAudioTrack);
```

Comment:
We should be able to update this to use the new `CONVERT_OR_THROW_AND_RETURN` macro.